### PR TITLE
T64 full bit-transpose variant

### PR DESCRIPTION
**Pull request by 4ertus2**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Category (leave one):
- Improvement

Short description (up to few sentences):
Full bit transpose variant for T64 codec. Could lead to better compression with zstd.

Detailed description (optional):
Continuation of #5557 and #5731. Allow to set option to T64 codec: T64('byte'), T64('bit'). So it would generate byte-transposed or bit-transposed data. By default it uses 'byte' variant.

### CODE REVIEW

**Comment by 4ertus2** [comment]
I've compressed some real data DateTime column with different variants:
```
lz4             833569732
zstd            515097904
t64(byte)+lz4   407128098
t64(bit)+lz4    472855596
t64(byte)+zstd  345330458
t64(bit)+zstd   318123008
```
It looks that full bit-transpose helps zstd and do not helps lz4.

**Comment by alexey-milovidov** [comment]
You can also compare with different levels of ZSTD, because level 1 is used by default.
When comparing by compressed size, you should also mention compression and decompression speed.

**Comment by alexey-milovidov** [comment]
PS. What's in stress test?

**Comment by 4ertus2** [comment]
> PS. What's in stress test?

Code: 202. DB::Exception: Received from localhost:9000, ::1. DB::Exception: Too many simultaneous queries. Maximum: 100.

**Comment by alexey-milovidov** [comment]
> Code: 202. DB::Exception: Received from localhost:9000, ::1. DB::Exception: Too many simultaneous queries. Maximum: 100.

This message cannot relate to stress test failure. It fails only in case of sanitizer reports or server crash.

File: dbms/src/Compression/CompressionCodecT64.cpp [hunk 1]
```diff
@@ -200,70 +228,64 @@ void transpose(const char * src, char * dst, UInt32 num_bits, UInt32 tail = 64)
...
+        UInt64 * mx_line = &mx[full_bytes * 8];
+        transpose64x8(mx_line);
+        memcpy(dst, mx_line, part_bits * sizeof(UInt64));
     }
 }
 
 /// UInt64[N] transposed matrix -> UIntX[64]
-template <typename _T, typename _MinMaxT = std::conditional_t<std::is_signed_v<_T>, Int64, UInt64>>
-void revTranspose(const char * src, char * dst, UInt32 num_bits, _MinMaxT min, _MinMaxT max [[maybe_unused]], UInt32 tail = 64)
+template <typename _T, bool _full = false>
```
Comment (alexey-milovidov) [inline dbms/src/Compression/CompressionCodecT64.cpp]: `_Capital` is reserved for C++ language and standard library.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
.

File: dbms/src/Compression/CompressionCodecT64.cpp (hunk 1 above)
Comment (alexey-milovidov) [reply dbms/src/Compression/CompressionCodecT64.cpp]: Let's remove leading underscores at all.

File: dbms/src/Compression/CompressionCodecT64.cpp [hunk 2]
```diff
@@ -200,70 +228,64 @@ void transpose(const char * src, char * dst, UInt32 num_bits, UInt32 tail = 64)
...
+void revTranspose(const char * src, _T * buf, UInt32 num_bits, UInt32 tail = 64)
 {
     UInt64 mx[64] = {};
```
Comment (alexey-milovidov) [inline dbms/src/Compression/CompressionCodecT64.cpp]: Is `mx` an abbreviation for `matrix`?
I hate abbreviations.

File: dbms/src/Compression/CompressionCodecT64.cpp [hunk 3]
```diff
@@ -200,70 +228,64 @@ void transpose(const char * src, char * dst, UInt32 num_bits, UInt32 tail = 64)
...
+            revTranspose64x8(mx_line);
     }
 
-    _T upper_min = 0;
-    if (num_bits < 64)
-        upper_min = UInt64(min) >> num_bits << num_bits;
+    if (part_bits)
+    {
+        UInt64 * mx_line = &mx[full_bytes * 8];
+        revTranspose64x8(mx_line);
```
Comment (alexey-milovidov) [inline dbms/src/Compression/CompressionCodecT64.cpp]: rev -> reverse

File: dbms/src/Compression/CompressionCodecT64.h [hunk 4]
```diff
@@ -12,12 +12,22 @@ class CompressionCodecT64 : public ICompressionCodec
     static constexpr UInt32 HEADER_SIZE = 1 + 2 * sizeof(UInt64);
     static constexpr UInt32 MAX_COMPRESSED_BLOCK_SIZE = sizeof(UInt64) * 64;
 
-    CompressionCodecT64(TypeIndex type_idx_)
+    enum class Variant
```
Comment (alexey-milovidov) [inline dbms/src/Compression/CompressionCodecT64.h]: Missing comment.

**Pull request merged**